src/hotspot/share/oops/methodData.cpp

```diff
@@ -1558,6 +1558,7 @@ void MethodData::print_value_on(outputStream* st) const {
 }
 
 void MethodData::print_data_on(outputStream* st) const {
+  MutexLocker ml(extra_data_lock(), Mutex::_no_safepoint_check_flag);
```

Comment:
Are we sure nothing calls this method when lock is already held? Safer to use `ConditionalMutexLocker` with self-locked check?

Reply:
I don't really like ConditionalMutexLocker because it feels like a hack when we don't understand the control flow, but that said, there was a case I found with ciMethodData printing that already has the lock.  It would have failed if any tests hit it.